R/utils.R

```diff
@@ -36,3 +36,85 @@ calc_disc_dist_quantile <- function(prob, days, quantile) {
   names(quantiles) <- as.character(quantile)
   quantiles
 }
+#' Creates a JSON version of the epiparameter database
+#'
+#' @description This converts the database from a data frame (read from the csv
+#' stored in the package) into a JSON data file and saves it. This functions
+#' allows the database to be edited and then the JSON file can be updated but
+#' calling this function.
+#'
+#' @param read_path A relative file path from the package root directory to read
+#' the parameters database from
+#' @param write_path A relative file path from the package root directory to
+#' write the JSON data file to
+#'
+#' @return Invisibly returns the JSON data
+#' @keywords internal
+make_json_data <- function(read_path = file.path("extdata", "parameters.csv"),
+                           write_path = file.path("inst", "extdata", "data.json")) {
+  # read in epiparameter database
+  data <- utils::read.csv(
+    file = system.file(
+      read_path,
+      package = "epiparameter",
+      mustWork = TRUE
+    ), header = TRUE
+  )
+
+  # convert cells with arrays to numeric
+  for (i in grep(pattern = "_ci$", x = colnames(data))) {
+    if (!all(is.na(data[[i]]))) {
+      data[[i]] <- lapply(
+        strsplit(x = data[[i]], split = ",", fixed = TRUE),
+        as.numeric
+      )
+    }
+
+  }
+
+  # convert data to json
+  data_json <- jsonlite::toJSON(data, dataframe = "columns", na = "null")
+  jsonlite::validate(data_json)
+
+  # save json data
+  jsonlite::write_json(data_json, path = write_path)
+
+  # invisibly return json data
+  invisible(data_json)
+}
+
+#' Creates a data schema in JSON format from a yaml file
+#'
+#' @description This converts the yaml data schema (data dictionary) stored in
+#' the package into a JSON schema and saves it. This functions allows the yaml
+#' to be the file which can be read and edited and then this file updates to
+#' the JSON schema to match the yaml schema.
+#'
+#' @param read_path A relative file path from the package root directory to read
+#' the yaml schema from
+#' @param write_path A relative file path from the package root directory to
+#' write the JSON schema to
+#'
+#' @return Invisibly returns the JSON schema
+#' @keywords internal
+make_json_schema <- function(read_path = file.path("extdata", "data_dictionary.yaml"),
+                             write_path = file.path("inst", "extdata", "schema.json")) {
+  # read schema
+  schema <- yaml::read_yaml(
+    system.file(
+      read_path,
+      package = "epiparameter",
+      mustWork = TRUE
+    )
+  )
+
+  # convert schema to json
+  schema_json <- jsonlite::toJSON(schema, auto_unbox = TRUE)
+  jsonlite::validate(schema_json)
+
+  # save json schema
+  jsonlite::write_json(schema_json, path = write_path)
+
+  # invisibly return json schema
+  invisible(schema_json)
+}
```

Comment:
`make_json_schema` and the saving of the data dictionary as a YAML file, which is then converted to JSON only for the purposes of validating the database, seems a bit convoluted. Is this a common workflow? I would suggest saving the JSON schema as a human-readable file as well, this makes it much easier to figure out changes when reviewing. That requires passing `pretty = TRUE` as well as `auto_unbox = TRUE` to get the `$schema` string.

Reply:
Seems resolved in https://github.com/epiverse-trace/epiparameter/commit/ffff318e5c0f595a2e4a3b6666832457f931a866

Reply:
An example of the workflow of going from yaml to JSON for validation can be found here: https://github.com/covid19-forecast-hub-europe/HubValidations